simpleline/render/renderer.py

```diff
@@ -0,0 +1,382 @@
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+# Author(s): Jiri Konecny <jkonecny@redhat.com>
+#
+
+import getpass
+import queue
+import sys
+import threading
+
+from simpleline.event_loop import ExitMainLoop
+from simpleline.event_loop.signals import ExceptionSignal, InputReadySignal, RenderScreenSignal, InputScreenSignal
+from simpleline.render import RendererUnexpectedError, INPUT_PROCESSED, INPUT_DISCARDED
+from simpleline.render.prompt import Prompt
+from simpleline.render.screen_stack import ScreenStack, ScreenData, ScreenStackEmptyException
+from simpleline.render.widgets import TextWidget
+
+RAW_INPUT_LOCK = threading.Lock()
+
+
+class Renderer(object):
+
+    def __init__(self, event_loop, renderer_stack=None):
+        """Constructor where you can pass your own renderer stack.
+
+        The ScreenStack will be used automatically if renderer stack will be None.
+
+        :param event_loop: Event loop used for the renderer.
+        :type event_loop: Class based on `simpleline.event_loop.AbstractEventLoop`.
+        :param renderer_stack: Input your own renderer stack if you need to.
+        :type renderer_stack: `simpleline.screen_stack.ScreenStack` based class.
+        """
+        self._quit_screen = None
+        self._quit_message = ""
+        self._event_loop = event_loop
+        self._width = 0
+        self._spacer = ""
+        self._input_error_counter = 0
+        self._input_thread = None
+        if renderer_stack:
+            self._screen_stack = renderer_stack
+        else:
+            self._screen_stack = ScreenStack()
+        self._register_handlers()
+
+        self.width = 80
+        self.redraw()
+
+    def _register_handlers(self):
+        self._event_loop.register_signal_handler(RenderScreenSignal, self._redraw_callback)
+        self._event_loop.register_signal_handler(InputScreenSignal, self._process_input_callback)
+
+    @property
+    def width(self):
+        return self._width
+
+    @width.setter
+    def width(self, width):
+        self._width = width
+        self._spacer = "\n".join(2 * [width * "="])
+
+    @property
+    def quit_screen(self):
+        return self._quit_screen
+
+    @quit_screen.setter
+    def quit_screen(self, quit_screen):
+        self._quit_screen = quit_screen
+
+    @property
+    def nothing_to_render(self):
+        """Is something for rendering in the renderer stack?
+
+        :return: True if the rendering stack is empty
+        :rtype: bool
+        """
+        return self._screen_stack.is_empty()
+
+    def schedule_screen(self, ui_screen, args=None):
```

Comment:
As noted earlier, but noting there as well for clarity: I think this should be `push_screen()`.